src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -314,17 +322,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T obj) where T : class
         {
-            if (diagnostic == null)
+            if (obj == null)
             {
                 Contract.Requires(false, "who returns invalid data?");
             }
         }
 
+        [Conditional("DEBUG")]
+        public void AssertCleanup()
+        {
+            foreach (var idAndData in _map.Values.ToList())
+            {
+                foreach (var data in idAndData.Values.ToList())
+                {
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
```

Comment:
> we need an aggregation service, since we currently don't have a service that can let one know all workspaces currently exist, we need CCW one way or the other.

I don't understand.  Can you please more thoroughly explain the design here.  Why do you need an "aggregation service"?  What is the "aggregation service" aggregating?   Why do you "need CCW one way or the other"?

> here map could be changed to per workspace CWT. but still need CWT.

Why do you need a CWT at all?  I cannot figure out from this code what purpose the CWT serves or why this code needs to hold onto a Workspace in the first place (either directly, or through a weak reference).  Why can't it just hold onto an ID?  What is the workspace needed for?